### ISSUE TITLE
Che Cluster Networking annotations only apply to ingresses, not routes

### ISSUE THREAD

**Issue by AObuchow** [eclipse/che#23118]
### Describe the bug

In the Che Cluster CR, there's a [field](https://github.com/eclipse-che/che-operator/blob/aa0150d74fe0db59f475192ba13230c68dfd7e72/api/v2/checluster_types.go#L244-L251) for setting annotations to be applied to the ingresses and routes created for workspaces:
```
	// Defines annotations which will be set for an Ingress (a route for OpenShift platform).
	// The defaults for kubernetes platforms are:
	//     kubernetes.io/ingress.class:                       "nginx"
	//     nginx.ingress.kubernetes.io/proxy-read-timeout:    "3600",
	//     nginx.ingress.kubernetes.io/proxy-connect-timeout: "3600",
	//     nginx.ingress.kubernetes.io/ssl-redirect:          "true"
	// +optional
	Annotations map[string]string `json:"annotations,omitempty"`
```

However, in the Che router implementation, these annotations are only applied to [ingresses](https://github.com/eclipse-che/che-operator/blob/main/controllers/devworkspace/solver/endpoint_exposer.go#L251-L253), not [routes](https://github.com/eclipse-che/che-operator/blob/main/controllers/devworkspace/solver/endpoint_exposer.go#L242-L247).

### Che version

7.90@latest

### Steps to reproduce

1. Configure `networking.annotations` in the Che Cluster CR to have a custom annotation, e.g. `my-test-annotation: test-annotation-value`
2. Create a workspace (e.g. the empty workspace sample)
3. Check the routes created for your workspace and see the custom annotations set in the Che Cluster CR is missing.

### Expected behavior

The annotations from `networking.annotations` in the Che Cluster CR are applied to the routes created for a workspace. 

### Runtime

OpenShift

### Installation method

OperatorHub

### Environment

Linux

**Comment by AObuchow** [comment]
It might be that this is simply an error in the Che Cluster CR's documentation. Maybe it is intended for the networking annotations to only get applied to Ingresses and not Routes.

**Comment by tolusha** [comment]
@AObuchow 
Can we close it ?

**Comment by batleforc** [comment]
Hi, is it possible to have another field in the CheCluster CRD that will put annotations automatically and only on the devfile's endpoint ?
We have a case where we need the default annotations on the dashboard but different one on the route or ingress that wouldn't be OPT-IN
And if possible to have this issue prioritized has it's would allow a different (with easier Cert management) in openshift

**Comment by AObuchow** [comment]
@batleforc does configuring `networking.Annotations` also set these annotations for the Dashboard?

**Comment by AObuchow** [comment]
@batleforc I tested setting `networking.Annotations` in the CheCluster (on minikube) and saw that they're also applied to an ingress called `eclipse-che` which _seems_ to be used by Che Server:

```YAML
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    che.eclipse.org/managed-annotations-digest: 7khu-6ssVoWp0zXcjyyLcsLS5TRHYyWps6-cBXhm-7I=
    nginx.ingress.kubernetes.io/proxy-connect-timeout: "3600"
    nginx.ingress.kubernetes.io/proxy-read-timeout: "3600"
    nginx.ingress.kubernetes.io/ssl-redirect: "true"
    test: test
  creationTimestamp: "2024-10-24T19:02:00Z"
  generation: 1
  labels:
    app.kubernetes.io/component: che
    app.kubernetes.io/instance: che
    app.kubernetes.io/managed-by: che-operator
    app.kubernetes.io/name: che
    app.kubernetes.io/part-of: che.eclipse.org
  name: che
  namespace: eclipse-che
  ownerReferences:
  - apiVersion: org.eclipse.che/v2
    blockOwnerDeletion: true
    controller: true
    kind: CheCluster
    name: eclipse-che
    uid: 3d51127e-6e14-4c66-92aa-f01dabab6d9e
  resourceVersion: "2683"
  uid: 4e0a0b7f-1461-4d75-ba67-39b07a6367c7
spec:
  ingressClassName: nginx
  rules:
  - host: 192.168.49.2.nip.io
    http:
      paths:
      - backend:
          service:
            name: che-gateway
            port:
              number: 8080
        path: /
        pathType: ImplementationSpecific
  tls:
  - hosts:
    - 192.168.49.2.nip.io
    secretName: che-tls
status:
  loadBalancer:
    ingress:
    - ip: 192.168.49.2
```

On [Kubernetes](https://github.com/eclipse-che/che-operator/blob/1daa6b0b834b07ca40be21e93323026b731dce13/pkg/deploy/server/chehost_reconciler.go#L72), an ingress will be created that seems to be used for the Che Server, and that ingress will use the [annotations](https://github.com/eclipse-che/che-operator/blob/1daa6b0b834b07ca40be21e93323026b731dce13/pkg/deploy/ingress.go#L96-L108) from `networking.Annotations` in the CheCluster CR.

Is this the problem you're describing?

**Comment by batleforc** [comment]
@AObuchow Yes, what I would love to prevent is having annotations for the main dashboard (One that work on the security and management) and have them also put on the end user's route that I don't want to monitor as an example.

On plain Kubernetes that's already the case, but on OpenShift that isn't and that's not bad, but I would love to be able to put default annotation on end user's route.

**Comment by AObuchow** [comment]
@batleforc sorry for the confusion on my end, I'm still not entirely sure I fully understand:
- Do you want the topic of this issue to be resolved? i.e.`networking.annotations` in the Che Cluster CR are also applied to the routes created for a workspace on OpenShift.
- Do you want a new field for annotations that **only apply to workspace routes/ingresses** (since it [seems](https://github.com/eclipse-che/che/issues/23118#issuecomment-2436160120) like the existing `networking.annotations` also applies to an ingress for che-server)
- Something else?

I'm not sure I see the connection with the main dashboard.

**Comment by batleforc** [comment]
@AObuchow , sorry i lost myself in the different need :/

I need the second on in reality it would allow more flexibility and possibly to apply a higher level of security on the dashboard

**Comment by batleforc** [comment]
Hey @AObuchow,
The annotations in the devfile's endpoint would have fixed the problem for openshift, if the che operator wouldn't erase the certificate injected by https://github.com/cert-manager/openshift-routes in the route.
For ingress, the fact that the certificate secret name is still the same as the other ingress (code-redirect for exemple) keep the ingress in error

**Comment by AObuchow** [comment]
Hey @batleforc, thank you for following up on this issue :) I'm sorry I never got to resolve it or the related issue https://github.com/eclipse-che/che/issues/22935.

Unfortunately, I'm no longer actively working on Eclipse Che so hopefully someone else from the Che/Dev Spaces team can take a look. CC: @ibuziuk @dkwon17

**Comment by batleforc** [comment]
@AObuchow Thanks for what you did till now for Che/Dev Spaces and related project, I will be happy to follow what you do in the future 👍

**Comment by AObuchow** [comment]
> @AObuchow Thanks for what you did till now for Che/Dev Spaces and related project, I will be happy to follow what you do in the future 👍 

@batleforc thank you, I appreciate it :) And thank you as well for all your contributions to Che & DWO, they are always greatly appreciated :)

**Comment by batleforc** [comment]
Hi @tolusha 
This issue is still there, we still have two problem. Each endpoint's related ingress (Tested on K3S and Talos) keep having the same secretName and for the routes (latest Openshift) when we use something like Cert Manager to inject the certificate one of the Che Operator kick him out.

**Comment by tolusha** [comment]
Hello @batleforc 
Do I understand correctly, you would like to have ability to inject its own certificate (via cert-manager for instance) into workspaces routes/ingresses. This will solve this issue and https://github.com/eclipse-che/che/issues/22935, right?

**Comment by batleforc** [comment]
Hi @tolusha 

Absolutely, its a little différent between route and ingress has route need to have the cert embeded inside the object (and once injected the che operator trigger a resolution and forcefully remove it) and the ingress need to refer to an external secret (that should differ from the other).

**Comment by tolusha** [comment]
Will all routes/ingresses in all user namespaces use the same certificate?
I'm considering adding the ability to configure a custom TLS secret, but if certificates differ between namespaces, it could complicate things.

**Comment by batleforc** [comment]
In all the use case that i have in mind, the certificate will never be the same and in the case of cert manager will be http01 issued

**Comment by tolusha** [comment]
What about a single workspace which consists of several endpoints. 
Will they have different certificates?

**Comment by batleforc** [comment]
Yes too, has they have different fqdn there will in most case be différent.
I don't want and cant have a wildcard certificate due to security reason.

**Comment by tolusha** [comment]
Thank you for the clarification, I will start working on that issue.

**Comment by tolusha** [comment]
@batleforc 
Will you have a chance to do a quick test of solution?

**Comment by batleforc** [comment]
I can in ~1hour

**Comment by tolusha** [comment]
I will prepare a draft PR tomorrow.

**Comment by tolusha** [comment]
For OpenShift
1. Create a CatalogSource
```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: eclipse-che-22935
  namespace: openshift-marketplace
spec:
  image: 'quay.io/eclipse/eclipse-che-olm-catalog:22935'
  sourceType: grpc
```

2. Deploy Eclipse Che from that catalogsource
3. Update CheCluster CR, add needed annotations/labels
```
spec:
  devEnvironments:
    networking:
      externalTLSConfig: true
      annotations:
         ...
      labels:
        ...
```

4. Start a workspace (existing running workspaces will not be affected)
5. Check routes for injected `TLS` config

https://github.com/eclipse-che/che-operator/pull/2025

**Comment by tolusha** [comment]
@batleforc 
Did you have a chance to test ?

**Comment by batleforc** [comment]
Hello, not yet some work took priority, i will do my max to test it soon

**Comment by batleforc** [comment]
Hello @tolusha, I've tryed it and it seems to work for Openshift's route.
I've tried:

- Create new workspaces
- Upgrade from local devfile
	- Switching from https to http a route and the other way
	- No change and the certificate didn't move

The only drawback possible is when i setup the tls annotations from the che CRD every route become HTTPs, i would love to have a way to have annotations for https route only.

**Comment by tolusha** [comment]
Hello. 

> I would love to have a way to have annotations for https route only.

Make sense!
Thank you for your feedback.

**Comment by tolusha** [comment]
@batleforc 
Is ready for review and testing https://github.com/eclipse-che/che-operator/pull/2025

**Comment by batleforc** [comment]
Hi @tolusha 
Thank's for the quick PR, just a question, in controllers/devworkspace/solver/endpoint_exposer.go line 298 you copy the TLS spec without changing the secret name, is it possible to append the endpoint name to the secret-name or is it already done and i didn't see it ?

**Comment by tolusha** [comment]
Please comment in the PR
